BPA/Ch.11/ipm_pred.stan

```diff
@@ -0,0 +1,228 @@
+// Integrated population model
+
+functions {
+  /**
+   * Return log probability of Poisson distribution.
+   * Outcome n may be a real value; for compatibility with OpenBUGS.
+   *
+   * @param n      Outcome
+   * @theta lambda Mean
+   *
+   * @return Log probability
+   */
+  real real_poisson_log(real n, real lambda) {
+    real lp;
+
+    if (lambda < 0) {
+      reject("lambda must be non-negative; found lambda=", lambda);
+    } else if (n < 0.0) {
+      reject("n must not be negative; found n=", n);
+    } else {
+      return n * log(lambda) - lambda - lgamma(n + 1);
+    }
+    return negative_infinity();
+  }
+
+  /**
+   * Return log probability of binomial distribution.
+   * Outcome n may be a real value; for compatibility with OpenBUGS.
+   *
+   * @param n     Outcome
+   * @param N     Size
+   * @theta theta Probability
+   *
+   * @return Log probability
+   */
+  real real_binomial_log(real n, real N, real theta) {
+    real lp;
+
+    if (N < 0) {
+      reject("N must be non-negative; found N=", N);
+    } else if (theta < 0 || theta > 1) {
+      reject("theta must be in [0,1]; found theta=", theta);
+    } else if (n < 0 || n > N) {
+      reject("n must be in [0:N]; found n=", n);
+    } else {
+      return binomial_coefficient_log(N, n)
+        + n * log(theta) + (N - n) * log(1.0 - theta);
+    }
+    return negative_infinity();
+  }
+}
+
+data {
+  int nyears;                    // Number of years
+  vector[nyears] y;              // Population counts
+  int J[nyears-1];               // Total number of nestings recorded
+  int R[nyears-1];               // Annual number of surveyed broods
+  int m[2*(nyears-1), nyears];   // m-array
+  int<lower=1> t_pred;           // Number of years to predict
+}
+
+parameters {
+  real<lower=0> sigma_y;           // Observation error
+  vector<lower=0>[nyears] N1;      // Number of 1-year juveniles
+  vector<lower=0>[nyears] Nad;     // Number of adults
+  real<lower=0,upper=1> mean_sjuv; // Mean survival prob. juveniles
+  real<lower=0,upper=1> mean_sad;  // Mean survival prob. adults
+  real<lower=0,upper=1> mean_p;    // Mean recapture prob.
+  real<lower=0> mean_fec;          // Mean productivity
+}
+
+transformed parameters {
+  vector<lower=0,upper=1>[nyears] sjuv;
+  vector<lower=0,upper=1>[nyears] sad;
+  vector<lower=0,upper=1>[nyears] p;
+  vector<lower=0>[nyears] f;
+  vector<lower=0>[nyears] Ntot;
+  simplex[nyears] pr[2*(nyears-1)];
+  vector<lower=0>[nyears-1] rho;
+
+  // Survival and recapture probabilities, as well as productivity
+  for (t in 1:nyears) {
+    sjuv[t] <- mean_sjuv;
+    sad[t] <- mean_sad;
+    p[t] <- mean_p;
+    f[t] <- mean_fec;
+  }
+
+  // Total number of individuals
+  for (t in 1:nyears)
+    Ntot[t] <- Nad[t] + N1[t];
+
+  { // m-array
```

Comment:
sometimes it can help readability if you break long blocks of code like this up into functions with informative names